compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -1626,8 +1626,26 @@ impl<'a> Parser<'a> {
     }
 
     /// Parses an optional `move` prefix to a closure-like construct.
-    fn parse_capture_clause(&mut self) -> CaptureBy {
-        if self.eat_keyword(kw::Move) { CaptureBy::Value } else { CaptureBy::Ref }
+    fn parse_capture_clause(&mut self) -> PResult<'a, CaptureBy> {
+        if self.eat_keyword(kw::Move) {
+            // Check for `move async` and recover
+            if self.check_keyword(kw::Async) {
+                let move_async_span = self.token.span.with_lo(self.prev_token.span.data().lo);
```

Comment:
Could you create a function in `compiler/rustc_parse/src/parser/diagnostics.rs` (e.g. `report_incorrect_move_async_order` or something like that) which reports this error and then call it here? 

It's a little bit nicer to keep the diagnostics together and avoid obscuring the parser logic (while this is only a small addition, these things sneak up on us).